README.md
# NimraylibNow! - The Ultimate Raylib wrapper for Nim

The most idiomatic and up-to-date wrapper for [Raylib] gaming C library.
Use this library if you want to write games using [Raylib] in [Nim].

This library is an effort to create an automatically generated wrapper which can
be easily upgraded to any future version of [Raylib]. Anyone should be able to
upgrade it with some effort and [HACKING](HACKING.md) should be a guide on how
to do it. Please file a bug report if any of that is too far from the reality.

[Nim]: https://nim-lang.org/

## Features

* Automated generation of the wrapper using the power of (((Nim)))

* Idiomatic Nim naming and conventions so you write **Nim** code, not C

* 60+ [examples](examples/) converted from C to Nim

* Includes modules: [raylib], [raymath], [rlgl], [raygui], [physac]

* Supports Windows, Linux, MacOS, Emscripten (see [Emscripten example])

* Supports static compilation, dynamic linking - you choose

[raylib]: https://github.com/raysan5/raylib
[raymath]: https://github.com/raysan5/raylib/blob/master/src/raymath.h
[rlgl]: https://github.com/raysan5/raylib/blob/master/src/rlgl.h
[raygui]: https://github.com/raysan5/raygui
[physac]: https://github.com/raysan5/raylib/blob/master/src/physac.h
[Emscripten example]: examples/emscripten/

## Quickstart

1. Install NimraylibNow!
```shell
$ nimble install nimraylib_now  # wait 10 minutes for download, sorry
```
2. Copy any example to your file, for instance [crown.nim](examples/original/crown.nim)
3. Compile and run it:
```shell
$ nim c -r -d:release --gc:orc crown.nim
```

Explanation of flags:
- `-r` - run after compiling
- `-d:release` - for speed, but also cuts away debugging information
- `--gc:orc` - memory management without garbage collection and
  with better C interoperability

## Current version and upgrading from previous versions

NimraylibNow! v0.15 is a wrapper for [Raylib v4.2.0] and [Raygui v3.2] and works
for Nim v1.4.2+.

See [CHANGELOG](CHANGELOG.md) for any tips on how to upgrade your code
from previous versions.

[Raylib v4.2.0]: https://github.com/raysan5/raylib/releases/tag/4.2.0
[Raygui v3.2]: https://github.com/raysan5/raygui/releases/tag/3.2

## Install NimraylibNow!

Install this library with nimble (takes **6-10(!) minutes** because of huge
original raylib repository):
```shell
$ nimble install nimraylib_now
```

Or put it into your `.nimble` file:
```nim
requires "nimraylib_now"
```

## How to use NimraylibNow!

Import any necessary modules and use it:

```nim
# Import everything but rlgl, should be enough for most cases!
import nimraylib_now

# Import only necessary modules for more granularity
import nimraylib_now/raylib
import nimraylib_now/[raylib, raymath, raygui]
from nimraylib_now/rlgl as rl import nil  # import rlgl with a mandatory prefix rl
```

See [examples](examples/) for how to use it.
You should generally be able to follow any tutorials for official C bindings,
just mind the [naming](#naming-differences-with-c). Also see official
[cheatsheet](https://www.raylib.com/cheatsheet/cheatsheet.html) and
directly inspect the binding sources, e.g. for
[raylib](src/nimraylib_now/not_mangled/raylib.nim).

Here is a long example to showcase most features
([crown.nim](examples/original/crown.nim)).

```nim
import math
import nimraylib_now

const
  nimFg: Color = (0xff, 0xc2, 0x00)          # Use this shortcut with alpha = 255!
  nimBg: Color = (0x17, 0x18, 0x1f)

# Let's draw a Nim crown!
const
  crownSides = 8                             # Low-polygon version
  centerAngle = 2.0 * PI / crownSides.float  # Angle from the center of a circle
  lowerRadius = 2.0                          # Lower crown circle
  upperRadius = lowerRadius * 1.4            # Upper crown circle
  mainHeight = lowerRadius * 0.8             # Height without teeth
  toothHeight = mainHeight * 1.3             # Height with teeth
  toothSkew = 1.2                            # Little angle for teeth

var
  lowerPoints, upperPoints: array[crownSides, tuple[x, y: float]]

# Get evenly spaced points on the lower and upper circles,
# use Nim's math module for that
for i in 0..<crownSides:
  let multiplier = i.float
  # Formulas are for 2D space, good enough for 3D since height is always same
  lowerPoints[i] = (
    x: lowerRadius * cos(centerAngle * multiplier),
    y: lowerRadius * sin(centerAngle * multiplier),
  )
  upperPoints[i] = (
    x: upperRadius * cos(centerAngle * multiplier),
    y: upperRadius * sin(centerAngle * multiplier),
  )

initWindow(800, 600, "[nim]RaylibNow!")  # Open window

var camera = Camera(
  position: (5.0, 8.0, 10.0),  # Camera position
  target: (0.0, 0.0, 0.0),     # Camera target it looks-at
  up: (0.0, 1.0, 0.0),         # Camera up vector (rotation over its axis)
  fovy: 45.0,                  # Camera field-of-view apperture in Y (degrees)
  projection: Perspective      # Defines projection type, see CameraProjection
)
camera.setCameraMode(Orbital)  # Several modes available, see CameraMode

var pause = false              # Pausing the game will stop animation

setTargetFPS(60)               # Set the game to run at 60 frames per second

# Wait for Esc key press or when the window is closed
while not windowShouldClose():
  if not pause:
    camera.addr.updateCamera   # Rotate camera

  if isKeyPressed(Space):      # Pressing Space will stop/resume animation
    pause = not pause

  beginDrawing:                # Use drawing functions inside this block
    clearBackground(RayWhite)  # Set background color

    beginMode3D(camera):       # Use 3D drawing functions inside this block
      drawGrid(10, 1.0)

      for i in 0..<crownSides:
        # Define 5 points:
        # - Current lower circle point
        # - Current upper circle point
        # - Next lower circle point
        # - Next upper circle point
        # - Point for peak of crown tooth
        let
          nexti = if i == crownSides - 1: 0 else: i + 1
          lowerCur: Vector3 = (lowerPoints[i].x, 0.0, lowerPoints[i].y)
          upperCur: Vector3 = (upperPoints[i].x, mainHeight, upperPoints[i].y)
          lowerNext: Vector3 = (lowerPoints[nexti].x, 0.0, lowerPoints[nexti].y)
          upperNext: Vector3 = (upperPoints[nexti].x, mainHeight, upperPoints[nexti].y)
          tooth: Vector3 = (
            (upperCur.x + upperNext.x) / 2.0 * toothSkew,
            toothHeight,
            (upperCur.z + upperNext.z) / 2.0 * toothSkew
          )

        # Front polygon (clockwise order)
        drawTriangle3D(lowerCur, upperCur, upperNext, nimFg)
        drawTriangle3D(lowerCur, upperNext, lowerNext, nimFg)

        # Back polygon (counter-clockwise order)
        drawTriangle3D(lowerCur, upperNext, upperCur, nimBg)
        drawTriangle3D(lowerCur, lowerNext, upperNext, nimBg)

        # Wire line for polygons
        drawLine3D(lowerCur, upperCur, Gray)

        # Crown tooth front triangle (clockwise order)
        drawTriangle3D(upperCur, tooth, upperNext, nimFg)

        # Crown tooth back triangle (counter-clockwise order)
        drawTriangle3D(upperNext, tooth, upperCur, nimBg)

    block text:
      block:
        let
          text = "I AM NIM"
          fontSize = 60
          textWidth = measureText(text, fontSize)
          verticalPos = (getScreenHeight().float * 0.4).int
        drawText(
          text,
          (getScreenWidth() - textWidth) div 2,  # center
          (getScreenHeight() + verticalPos) div 2,
          fontSize,
          Black
        )
      block:
        let text =
          if pause: "Press Space to continue"
          else: "Press Space to pause"
        drawText(text, 10, 10, 20, Black)

closeWindow()
```

![Nim crown drawn with polygons](crown.png?raw=true)

## Naming differences with C
Naming is converted to more Nim conventional style but in a predictable manner.
Generally just omit any prefixes you see in official docs and use camelCase for
procs, everything else stays the same:
```c
void InitWindow(int width, int height, const char *title);
void GuiSetState(int state);
void rlBegin(int mode);
```
to
```nim
proc initWindow*(width: cint; height: cint; title: cstring)
proc setState*(state: cint)
proc begin*(mode: cint)
```

Although some definitions may have names like `FULLSCREEN_MODE`, you can still
use them (and encouraged to) as `FullscreenMode` in code.

### Raymath

All prefixes specific to types were omitted where possible:
```c
Vector2 Vector2Zero(void);
Vector3 Vector3Zero(void);

Vector2 Vector2Add(Vector2 v1, Vector2 v2);
Vector3 Vector3Add(Vector3 v1, Vector3 v2);
```
to
```nim
proc vector2Zero*(): Vector2
proc vector3Zero*(): Vector3

proc add*(v1: Vector2; v2: Vector2): Vector2
proc add*(v1: Vector3; v2: Vector3): Vector3
```

## Passing values by addresses
**BE VERY CAREFUL** with values which are passed as addresses, **always** convert
them to approapriate C-compatible types:
```nim
# Will do incorrect things!
var
  xPos = 0.0  # this is `float` type, will not be converted to `cfloat` automatically
  cameraPos = [xPos, camera.position.y, camera.position.z]
setShaderValue(shader, viewEyeLoc, cameraPos.addr, Vec3)  # cameraPos.addr is a pointer

# Convert values instead
var xPos: cfloat = 0.0
# or
var cameraPos = [xPos.cfloat, camera.position.y, camera.position.z]
# So this finally works as intended:
setShaderValue(shader, viewEyeLoc, cameraPos.addr, Vec3)
```

## Tips and tricks
### Tuple to object converters for geometry
Vector2, Vector3, Vector4 (Quaternion), Matrix, Rectangle, Color can be written
as a tuple to save on typing as they have pretty much standard parameter
sequence:

```nim
# All are valid:
var
  v1: Vector2 = (x: 3.0, y: 5.0)
  v2 = Vector2(x: 3.0, y: 5.0)
  v3 = (3.0, 5.0)
  c: Color = (0xc9, 0xc9, 0xc9)  # color can be written as a tuple even without the alpha value!
```

### Require fully qualified procs
For functions operating on global scope it could be convenient to use
fully qualified proc name:
```nim
rlgl.begin(rlgl.Triangles)
# draw something
rlgl.end()
```
which can be enforced by the compiler by importing `rlgl` module like this:
```nim
from nimraylib_now/rlgl import nil

# or to use a shorter `rl`
from nimraylib_now/rlgl as rl import nil
```

### Begin-End pairs sugar
There are pairs like `beginDrawing()` - `endDrawing()`, each have helping
templates to automatically insert `end`-proc at the end:
```nim
beginDrawing:
  drawLine(...)
```
is same as
```nim
beginDrawing()
drawLine(...)
endDrawing()
```

### Reserved words
Be careful and use "stropping" when using fields or functions which are also
reserved words in Nim:
```nim
`end`()
```

### Enums
All enums are marked as `{.pure.}` which means they should be fully qualified
when compiler can't guess their type:
```nim
camera.projection = Perspective  # can be guessed, no need for CameraProjection.Perspective
if isKeyDown(Right):             # can be guessed, no need for KeyboardKey.Right
if KeyboardKey.Right.isKeyDown:  # cannot be guessed
```

### Raymath infix operators
You can use infix operators like `+`, `-`, `*`, `/` for operations on vectors,
matrices and quaternions:
```nim
let
  v1 = Vector3(x: 5, y: 2, z: 1)
  v2 = Vector3(x: 3, y: 0, z: 0)
assert v1 - v2 == Vector3(x: 2.0, y: 2.0, z: 1.0)
```

## Raylib, the original C library

If you would like to know more about how this wrapper uses [Raylib] library
and see all the different options, check [USING_RAYLIB](/USING_RAYLIB.md),
it should help.

## Contributing

Do you want to contribute but don't know how? Check out
[CONTRIBUTING](CONTRIBUTING.md)!

## How does wrapper work?

If you would like to know how this wrapper works or how to update it, check
out [HACKING](HACKING.md).

## Troubleshooting
### Freezes on Wayland
Random freezes when everything stops being responsive can be caused by glfw
compiled against Wayland library. Try X11 version of glfw.

[GLFW] is a raylib dependency that was installed by your package manager, you
can uninstall raylib with clearing all of its dependencies and install it again
choosing X11 version of glfw.

[GLFW]: https://www.glfw.org/

### Windows error: The application was unable to start correctly (0x00000007b)
Or similar errors which point to missing `libwinpthread-1.dll` file mean that
windows can't find correct library files.

This is because, by default, nim creates binaries that are dynamically linked.
Normally, this is fine since windows already has the libraries required for most
basic nim applications. It does not however have libwinpthread-1.dll, which causes an error here.
We can solve this by instructing nim to create binaries with the libraries
linked statically.

To do this just pass in the `--passL:-static` to nim when compiling your file.

If you still want dynamic linking:

- If MinGW is not in PATH
- Add Nim's minGW toolchain to your PATH (You have to add the bin directory with .dll files not the one where nim.exe is located)

If you already have a version of MinGW (that is already on PATH and seperate from Nim)
- Copy over `libwinpthread-1.dll` from nim's version of MinGW into the bin directory of your version of MinGW

## Community

Find us at `#raylib-nim` channel
in the official [Raylib discord server](https://discord.gg/raylib).

## Thanks

Many thanks to V.A. Guevara for the efforts on [Raylib-Forever] library which
was the initial inspiration for this library.

[Raylib-Forever]: https://github.com/Guevara-chan/Raylib-Forever

Thanks to everyone who contributed to this project!

## License

MIT licensed, see LICENSE.md.


CHANGELOG.md
# Changelog

This guide documents changes between different versions of NimraylibNow!
and and if necessary simultaneously different versions of Raylib library.

There are not a lot of possibilities to use proper deprecation warnings
since it is not always possible to translate `#define` declarations from C.
Sadly upgrades of user code is a necessary part of migration. On the bright
side it should mainly consist of renaming, please try the wonderful tool
`nimgrep`, it is shipped with Nim and has a useful "replace" functionality.

Additionally you can consult [Raylib's changelog](https://github.com/raysan5/raylib/blob/master/CHANGELOG).

## NimraylibNow! v0.15, Raylib v4.2 and Raygui v3.2

NimraylibNow!-specific changes:
- Emscripten support is improved: added TOTAL_MEMORY option and it now allows
  using resources from your file system. See [Emscripten example readme](/examples/emscripten/README.md),
  "Compile options" section.

Raylib-specific changes - consult [Raylib's changelog](https://github.com/raysan5/raylib/blob/master/CHANGELOG) file.

## NimraylibNow! v0.14 and Raylib v4.0

NimraylibNow!-specific changes:
- rlgl module has a lot of intersecting identifiers with raylib module,
  you will have to fully specify some variables when both modules are
  imported.
- Mouse buttons and some keyboard keys will have to be fully specified,
  `MouseButton.Left` instead of `LeftButton` since now they intersect
  with keyboard buttons.

Raylib-specific changes - consult [Raylib's changelog](https://github.com/raysan5/raylib/blob/master/CHANGELOG) file.

## From NimraylibNow! v0.12 to v0.13

- `Mesh` type has most of its fields changed from `ptr cfloat` to `ptr UncheckedArray[cfloat]`
- `MouseCursor` enum has no `CURSOR_` prefix, e.g. `CURSOR_ARROW` is now `ARROW`
- `GamepadButton` enum has no `BUTTON_` prefix e.g. `BUTTON_LEFT_FACE_UP` is now `LEFT_FACE_UP`
- `GamepadAxis` enum has no `AXIS_` prefix, e.g. `AXIS_LEFT_X` is now `LEFT_X`
- `raygui_Support_Icons` define was removed from Raygui, it is enabled by default

## From Raylib v3.5 to v3.7 and from NimraylibNow! v0.10 to v0.12

Recommended:

- Change all imports except for `rlgl` to just `import nimraylib_now`

Renamed:

- `Camera.type` was renamed to `Camera.projection`
- `CameraType` was renamed to `CameraProjection`
- `GestureType` was renamed to `Gestures`
- `Material.params` type was changed from `ptr cfloat` to `array[4, cfloat]`
- `ConfigFlag` was renamed to `ConfigFlags`
- `TraceLogType` was renamed to `TraceLogLevel`
- `TextureFilterMod` was renamed to `TextureFilter`
- `TextureWrapMode` was renamed to `TextureWrap`
- `CubemapLayoutType` was renamed to `CubemapLayout`
- `NPatchInfo.type` was renamed to `NPatchInfo.layout`
- `NPatchType` was renamed to `NPatchLayout`
- `NPatchLayout` enum values were renamed to use English words instead of numbers
- `beginVrDrawing` was renamed to `beginVrStereoMode`
- `runPhysicsStep` was renamed to `updatePhysics`
- `getShaderDefault` and freinds were moved from `raylib` module to `rlgl`

Misc changes:

- VR handling was changed significantly, see updated example `core/core_vr_simulator.nim`
- Importing just `nimraylib_now` is now possible and it imports all libraries
  but `rlgl`
- "Tuple to object converters for geometry" feature requires importing
  `nimraylib_now/converters` or just `nimraylib_now`


raylib_gen.nim
import common, std/[streams, strutils]

const
  extraTypes = {
    "Vector4": "Quaternion* = Vector4 ## Quaternion, 4 components (Vector4 alias)",
    "Texture": "Texture2D* = Texture ## Texture2D, same as Texture",
    "Texture": "TextureCubemap* = Texture ## TextureCubemap, same as Texture",
    "RenderTexture": "RenderTexture2D* = RenderTexture ## RenderTexture2D, same as RenderTexture",
    "Camera3D": "Camera* = Camera3D ## Camera type fallback, defaults to Camera3D",
    "AudioStream": """RAudioBuffer* {.importc: "rAudioBuffer", header: "raylib.h", bycopy.} = object"""
  }
  raylibHeader = """
{.passL: "-lraylib -lGL -lm -lpthread -ldl -lrt -lX11 -DPLATFORM_DESKTOP".}

const
  RaylibVersion* = "4.1-dev"

type va_list* {.importc: "va_list", header: "<stdarg.h>".} = object ## Only used by TraceLogCallback
proc vprintf*(format: cstring, args: va_list) {.cdecl, importc: "vprintf", header: "<stdio.h>".}

## Callbacks to hook some internal functions
## WARNING: This callbacks are intended for advance users
type
  TraceLogCallback* = proc (logLevel: cint; text: cstring; args: va_list) {.cdecl.} ## Logging: Redirect trace log messages
  LoadFileDataCallback* = proc (fileName: cstring; bytesRead: ptr uint32): ptr UncheckedArray[uint8] {.
      cdecl.} ## FileIO: Load binary data
  SaveFileDataCallback* = proc (fileName: cstring; data: pointer; bytesToWrite: uint32): bool {.
      cdecl.} ## FileIO: Save binary data
  LoadFileTextCallback* = proc (fileName: cstring): cstring {.cdecl.} ## FileIO: Load text data
  SaveFileTextCallback* = proc (fileName: cstring; text: cstring): bool {.cdecl.} ## FileIO: Save text data
"""
  extraDistinct = """

  MaterialMapDiffuse* = MaterialMapAlbedo
  MaterialMapSpecular* = MaterialMapMetalness

  ShaderLocMapDiffuse* = ShaderLocMapAlbedo
  ShaderLocMapSpecular* = ShaderLocMapMetalness
  # Taken from raylib/src/config.h
  MaxShaderLocations* = ShaderLocationIndex(32) ## Maximum number of shader locations supported
  MaxMaterialMaps* = MaterialMapIndex(12) ## Maximum number of shader maps supported
  MaxMeshVertexBuffers* = 7 ## Maximum vertex buffers (VBO) per mesh

type
  Enums = ConfigFlags|Gesture
  Flag*[E: Enums] = distinct uint32

proc flag*[E: Enums](e: varargs[E]): Flag[E] {.inline.} =
  var res = 0'u32
  for val in items(e):
    res = res or uint32(val)
  Flag[E](res)
"""
  helpers = """

const
  LightGray* = Color(r: 200, g: 200, b: 200, a: 255)
  Gray* = Color(r: 130, g: 130, b: 130, a: 255)
  DarkGray* = Color(r: 80, g: 80, b: 80, a: 255)
  Yellow* = Color(r: 253, g: 249, b: 0, a: 255)
  Gold* = Color(r: 255, g: 203, b: 0, a: 255)
  Orange* = Color(r: 255, g: 161, b: 0, a: 255)
  Pink* = Color(r: 255, g: 109, b: 194, a: 255)
  Red* = Color(r: 230, g: 41, b: 55, a: 255)
  Maroon* = Color(r: 190, g: 33, b: 55, a: 255)
  Green* = Color(r: 0, g: 228, b: 48, a: 255)
  Lime* = Color(r: 0, g: 158, b: 47, a: 255)
  DarkGreen* = Color(r: 0, g: 117, b: 44, a: 255)
  SkyBlue* = Color(r: 102, g: 191, b: 255, a: 255)
  Blue* = Color(r: 0, g: 121, b: 241, a: 255)
  DarkBlue* = Color(r: 0, g: 82, b: 172, a: 255)
  Purple* = Color(r: 200, g: 122, b: 255, a: 255)
  Violet* = Color(r: 135, g: 60, b: 190, a: 255)
  DarkPurple* = Color(r: 112, g: 31, b: 126, a: 255)
  Beige* = Color(r: 211, g: 176, b: 131, a: 255)
  Brown* = Color(r: 127, g: 106, b: 79, a: 255)
  DarkBrown* = Color(r: 76, g: 63, b: 47, a: 255)
  White* = Color(r: 255, g: 255, b: 255, a: 255)
  Black* = Color(r: 0, g: 0, b: 0, a: 255)
  Blank* = Color(r: 0, g: 0, b: 0, a: 0)
  Magenta* = Color(r: 255, g: 0, b: 255, a: 255)
  RayWhite* = Color(r: 245, g: 245, b: 245, a: 255)
"""
  enumInFuncReturn = [
    ("GetKeyPressed", 0),
    ("GetGamepadButtonPressed", 14),
    ("GetGestureDetected", 25)
  ]
  enumInFuncParams = [
    # KeyboardKey
    ("IsKeyPressed", "key"),
    ("IsKeyDown", "key"),
    ("IsKeyReleased", "key"),
    ("IsKeyUp", "key"),
    ("SetExitKey", "key"),
    ("SetCameraPanControl", "keyPan"),
    ("SetCameraAltControl", "keyAlt"),
    ("SetCameraSmoothZoomControl", "keySmoothZoom"),
    ("SetCameraMoveControls", "keyFront"),
    ("SetCameraMoveControls", "keyBack"),
    ("SetCameraMoveControls", "keyRight"),
    ("SetCameraMoveControls", "keyLeft"),
    ("SetCameraMoveControls", "keyUp"),
    ("SetCameraMoveControls", "keyDown"),
    # GamepadButton
    ("IsGamepadButtonPressed", "button"),
    ("IsGamepadButtonDown", "button"),
    ("IsGamepadButtonReleased", "button"),
    ("IsGamepadButtonUp", "button"),
    # GamepadAxis
    ("GetGamepadAxisMovement", "axis"),
    # MouseCursor
    ("SetMouseCursor", "cursor"),
    # MouseButton
    ("IsMouseButtonPressed", "button"),
    ("IsMouseButtonDown", "button"),
    ("IsMouseButtonReleased", "button"),
    ("IsMouseButtonUp", "button"),
    # Gesture
    ("SetGesturesEnabled", "flags"),
    ("IsGestureDetected", "gesture"),
    # ConfigFlags
    ("SetConfigFlags", "flags"),
    ("SetWindowState", "flags"),
    ("ClearWindowState", "flags"),
    ("IsWindowState", "flag"),
    # TraceLogLevel
    ("TraceLog", "logLevel"),
    ("SetTraceLogLevel", "logLevel"),
    # CameraMode
    ("SetCameraMode", "mode"),
    # BlendMode
    ("BeginBlendMode", "mode"),
    # MaterialMapIndex
    ("SetMaterialTexture", "mapType"),
    # ShaderUniformDataType
    ("SetShaderValue", "uniformType"),
    ("SetShaderValueV", "uniformType"),
    # PixelFormat
    ("LoadImageRaw", "format"),
    ("ImageFormat", "newFormat"),
    ("GetPixelColor", "format"),
    ("SetPixelColor", "format"),
    ("GetPixelDataSize", "format"),
    # TextureFilter
    ("SetTextureFilter", "filter"),
    # TextureWrap
    ("SetTextureWrap", "wrap"),
    # CubemapLayout
    ("LoadTextureCubemap", "layout"),
    # FontType
    ("LoadFontData", "type")
  ]
  enumInFuncs = [
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "KeyboardKey",
    "GamepadButton",
    "GamepadButton",
    "GamepadButton",
    "GamepadButton",
    "GamepadAxis",
    "MouseCursor",
    "MouseButton",
    "MouseButton",
    "MouseButton",
    "MouseButton",
    "Flag[Gesture]",
    "Gesture",
    "Flag[ConfigFlags]",
    "Flag[ConfigFlags]",
    "Flag[ConfigFlags]",
    "ConfigFlags",
    "TraceLogLevel",
    "TraceLogLevel",
    "CameraMode",
    "BlendMode",
    "MaterialMapIndex",
    "ShaderUniformDataType",
    "ShaderUniformDataType",
    "PixelFormat",
    "PixelFormat",
    "PixelFormat",
    "PixelFormat",
    "PixelFormat",
    "TextureFilter",
    "TextureWrap",
    "CubemapLayout",
    "FontType"
  ]
  excludedFuncs = [
    # Text strings management functions
    "TextCopy",
    "TextIsEqual",
    "TextLength",
    "TextFormat",
    "TextSubtext",
    "TextReplace",
    "TextInsert",
    "TextJoin",
    "TextSplit",
    "TextAppend",
    "TextFindIndex",
    "TextToUpper",
    "TextToLower",
    "TextToPascal",
    "TextToInteger",
    # Misc
    "GetRandomValue",
    "SetRandomSeed",
    "OpenURL",
    # Files management functions
    "LoadFileData",
    "UnloadFileData",
    "SaveFileData",
    "LoadFileText",
    "UnloadFileText",
    "SaveFileText",
    "FileExists",
    "DirectoryExists",
    "IsFileExtension",
    "GetFileExtension",
    "GetFileName",
    "GetFileNameWithoutExt",
    "GetDirectoryPath",
    "GetPrevDirectoryPath",
    "GetWorkingDirectory",
    "GetDirectoryFiles",
    "ClearDirectoryFiles",
    "ChangeDirectory",
    "GetFileModTime",
    # Compression/Encoding functionality
    "CompressData",
    "DecompressData",
    "EncodeDataBase64",
    "DecodeDataBase64",
    # Text codepoints management functions (unicode characters)
    "LoadCodepoints",
    "UnloadCodepoints",
    "GetCodepointCount",
    "GetCodepoint",
    "CodepointToUTF8",
    "TextCodepointsToUTF8",
  ]
  allocFuncs = [
    "MemAlloc",
    "MemRealloc",
    "MemFree"
  ]
  privateFuncs = [
    "GetMonitorName",
    "GetClipboardText",
    "GetDroppedFiles",
    "GetGamepadName",
    "LoadModelAnimations",
    "UnloadModelAnimations",
    "LoadWaveSamples",
    "UnloadWaveSamples",
    "LoadImagePalette",
    "UnloadImagePalette",
    "LoadImageColors",
    "UnloadImageColors",
    "LoadFontData",
    "UnloadFontData",
    "LoadMaterials",
    "DrawLineStrip",
    "DrawTriangleFan",
    "DrawTriangleStrip",
    "LoadImageFromMemory",
    "DrawTexturePoly",
    "LoadFontEx",
    "LoadFontFromMemory",
    "GenImageFontAtlas",
    "DrawTriangleStrip3D",
    "DrawMeshInstanced",
    "LoadWaveFromMemory",
    "LoadMusicStreamFromMemory"
  ]

proc getReplacement(x, y: string, replacements: openarray[(string, string, string)]): string =
  # Manual replacements for some fields
  result = ""
  for a, b, pattern in replacements.items:
    if x == a and y == b:
      return pattern

proc genBindings(t: Topmost, fname: string; header, middle, footer: string) =
  var buf = newStringOfCap(50)
  var indent = 0
  var otp: FileStream
  try:
    otp = openFileStream(fname, fmWrite)
    lit header
    # Generate enum definitions
    lit "\ntype"
    scope:
      for enm in items(t.enums):
        spaces
        ident enm.name
        lit "* = distinct int32"
        doc enm
    lit "\n\nconst"
    scope:
      for enm in items(t.enums):
        for i, val in pairs(enm.values):
          spaces
          ident camelCaseAscii(val.name)
          lit "* = "
          ident enm.name
          lit "("
          lit $val.value
          lit ")"
          doc val
        lit "\n"
    lit extraDistinct
    # Generate type definitions
    lit "\ntype"
    scope:
      for obj in items(t.structs):
        spaces
        ident obj.name
        lit "* {.header: \"raylib.h\", bycopy.} = object"
        doc obj
        scope:
          for fld in items(obj.fields):
            spaces
            var (name, pat) = transFieldName(fld.name)
            ident name
            lit "*: "
            const replacements = [
              ("Camera3D", "projection", "CameraProjection"),
              ("Image", "format", "PixelFormat"),
              ("Texture", "format", "PixelFormat"),
              ("NPatchInfo", "layout", "NPatchLayout")
            ]
            let kind = getReplacement(obj.name, name, replacements)
            if kind != "":
              lit kind
            else:
              let many = isPlural(name) or (obj.name, name) in {"Model": "meshMaterial", "Model": "bindPose"}
              const replacements = [
                ("ModelAnimation", "framePoses", "ptr UncheckedArray[ptr UncheckedArray[$1]]"),
                ("Mesh", "vboId", "ptr array[MaxMeshVertexBuffers, $1]"),
                ("Material", "maps", "ptr array[MaxMaterialMaps, $1]"),
                ("Shader", "locs", "ptr array[MaxShaderLocations, $1]")
              ]
              let tmp = getReplacement(obj.name, name, replacements)
              if tmp != "": pat = tmp
              let kind = convertType(fld.`type`, pat, many, false)
              lit kind
            doc fld
        # Add a type alias or a missing type after the respective type.
        for name, extra in extraTypes.items:
          if obj.name == name:
            spaces
            lit extra
        lit "\n"
    lit middle
    # Generate functions
    lit "\n{.push callconv: cdecl, header: \"raylib.h\".}"
    for fnc in items(t.functions):
      if fnc.name in excludedFuncs: continue
      lit "\nproc "
      ident uncapitalizeAscii(fnc.name) # Follow Nim's naming convention for proc names.
      let isPrivate = fnc.name in privateFuncs
      let isAlloc = fnc.name in allocFuncs
      if isPrivate:
        lit "Priv("
      elif isAlloc:
        lit "("
      else:
        lit "*("
      var hasVarargs = false
      for i, (param, kind) in fnc.params.pairs:
        if param == "" and kind == "": # , ...) {
          hasVarargs = true
        else:
          if i > 0: lit ", "
          ident param
          lit ": "
          block outer:
            for j, (name, param1) in enumInFuncParams.pairs:
              if name == fnc.name and param1 == param:
                lit enumInFuncs[j]
                break outer
            let many = (fnc.name, param) != ("LoadImageAnim", "frames") and isPlural(param)
            const
              replacements = [
                ("GenImageFontAtlas", "recs", "ptr ptr UncheckedArray[$1]")
              ]
            let pat = getReplacement(fnc.name, param, replacements)
            let kind = convertType(kind, pat, many, not isPrivate)
            lit kind
      lit ")"
      if fnc.returnType != "void":
        lit ": "
        block outer:
          for (name, idx) in enumInFuncReturn.items:
            if name == fnc.name:
              lit enumInFuncs[idx]
              break outer
          let many = isPlural(fnc.name) or fnc.name == "LoadImagePalette"
          let kind = convertType(fnc.returnType, "", many, not isPrivate)
          lit kind
      lit " {.importc: \""
      ident fnc.name
      lit "\""
      if hasVarargs:
        lit ", varargs"
      lit ".}"
      if not (isAlloc or isPrivate) and fnc.description != "":
        scope:
          spaces
          lit "## "
          lit fnc.description
    lit "\n{.pop.}\n"
    lit footer
  finally:
    if otp != nil: otp.close()

const
  raylibApi = "../api/raylib_api.json"
  outputname = "../raylib.nim"

proc main =
  var t = parseApi(raylibApi)
  genBindings(t, outputname, raylibHeader, helpers, readFile("raylib_wrap.nim"))

main()
